File: beacon/ble_scan.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# ble_scan.rb — passive BLE presence counter for foot-traffic testing
# ---------------------------------------------------------------------
# Pure Ruby stdlib. Does NOT use a native BLE library — it drives BlueZ's
# own `bluetoothctl` and parses the stream. Works on Raspberry Pi OS.
#
# Setup (one time):
#   sudo apt update && sudo apt install -y ruby
#
# Run:
#   ruby ble_scan.rb                 # logs to ble_log_<timestamp>.csv
#   ruby ble_scan.rb mylog.csv       # custom file
#   Ctrl-C to stop -> prints a summary.
#
# If you get a permissions/"No default controller" error, either run with
# sudo, or add yourself to the bluetooth group:  sudo usermod -aG bluetooth $USER
# (then log out/in), and make sure the service is up:
#   sudo systemctl enable --now bluetooth
#
# IMPORTANT honesty note: this counts DISTINCT MAC ADDRESSES seen. Modern
# phones randomize their BLE MAC, so one phone can appear as several MACs.
# Treat the unique count as an UPPER BOUND, not a true person count. The
# point of this test is to learn whether passing phones show up AT ALL and
# how strong the signal is.

require "csv"
require "time"

csv_path = ARGV[0] || "ble_log_#{Time.now.strftime('%Y%m%d_%H%M%S')}.csv"

MAC_RE  = /([0-9A-F]{2}(?::[0-9A-F]{2}){5})/i
RSSI_RE = /RSSI:\s*(-?\d+)/i
# property keywords that follow a MAC but are NOT a device name
PROP_RE = /\A(RSSI|UUIDs?|ManufacturerData|ServiceData|TxPower|Connected|Paired|
              Trusted|Bonded|LegacyPairing|Modalias|Appearance|Icon|Class|
              ServicesResolved|WakeAllowed|AdvertisingFlags|AdvertisingData|
              Adapter|Alias|Blocked)\b/xi

seen = {} # mac => { first:, last:, hits:, rssi:, name: }

csv = CSV.open(csv_path, "w")
csv << %w[timestamp event mac rssi name unique_so_far]
csv.flush

puts "Logging to #{csv_path}"
puts "Starting BLE scan via bluetoothctl ... (Ctrl-C to stop)"
puts "-" * 64

at_exit do
  csv.flush rescue nil
  csv.close rescue nil
  puts "\n" + "=" * 64
  puts "SCAN SUMMARY"
  puts "  Unique MAC addresses seen : #{seen.size}"
  if seen.any?
    secs = (Time.now - seen.values.map { |v| v[:first] }.min).round
    puts "  Scan duration             : #{secs} sec"
    strongest = seen.values.map { |v| v[:rssi] }.compact.max
    puts "  Strongest RSSI (closest)  : #{strongest} dBm" if strongest
  end
  puts "  CSV saved to              : #{csv_path}"
  puts "  Reminder: count is an UPPER BOUND (phones rotate MACs)."
  puts "=" * 64
end

def strip_ansi(line)
  line.gsub(/\e\[[0-9;?]*[ -\/]*[@-~]/, "")
end

# Drive bluetoothctl with line-buffered output so we see devices live.
IO.popen(["stdbuf", "-oL", "bluetoothctl"], "r+") do |bt|
  bt.puts "power on"
  bt.puts "scan on"

  bt.each_line do |raw|
    line = strip_ansi(raw)
    m = line.match(MAC_RE)
    next unless m
    mac = m[1].upcase

    rssi = line[RSSI_RE, 1]&.to_i

    name = line[/Device\s+#{Regexp.escape(mac)}\s+(.+?)\s*\z/i, 1]
    name = nil if name && name.match?(PROP_RE)

    now = Time.now
    if (rec = seen[mac])
      rec[:last] = now
      rec[:hits] += 1
      rec[:rssi] = rssi if rssi
      rec[:name] = name if name && !name.empty?
      csv << [now.iso8601, "seen", mac, rssi, rec[:name], seen.size]
    else
      seen[mac] = { first: now, last: now, hits: 1, rssi: rssi, name: name }
      csv << [now.iso8601, "NEW", mac, rssi, name, seen.size]
      csv.flush
      printf("[%s] NEW  %s  RSSI=%-4s %-20s (unique: %d)\n",
             now.strftime("%H:%M:%S"), mac, (rssi || "?").to_s, name.to_s, seen.size)
    end
  end
end

File: beacon/ble_scan.py
#!/usr/bin/env python3
"""
ble_scan.py - passive BLE presence counter for foot-traffic testing.

Uses bleak (>=3.0), which talks to BlueZ over D-Bus (the real API - robust for
long, unattended runs, unlike scraping bluetoothctl text). Uses the modern
async-iterator API `scanner.advertisement_data()` (bleak 3.x).

------------------------------------------------------------------------------
SETUP on Raspberry Pi OS (tested on Debian trixie / Python 3.13 on a Pi Zero 2 W;
also works on Bookworm). Python 3 is preinstalled:
    sudo apt update && sudo apt install -y python3-venv
    python3 -m venv ~/ble-env
    source ~/ble-env/bin/activate
    pip install bleak

RUN (with the venv active):
    python ble_scan.py                 # logs to ble_log_<timestamp>.csv
    python ble_scan.py mylog.csv       # custom filename
    Ctrl-C to stop -> prints a summary.

If bluetoothctl show reports "Powered: no / PowerState: off-blocked", the radio
is rfkill-blocked. Enable it once:
    sudo rfkill unblock bluetooth
    bluetoothctl power on
No sudo is needed for scanning itself.
------------------------------------------------------------------------------
FIELD FINDINGS (Jul 10 2026, Pi Zero 2 W "bob1"):
  * Phones ARE detectable - a phone brought near appeared at strong RSSI
    (~-50 dBm). Passive BLE sensing of phones works.
  * BUT one phone = MULTIPLE rotating MACs. Toggling a phone's Bluetooth
    produced 2+ new "devices" at once (Resolvable Private Addresses). Phones
    advertise several services on separate, rotating private addresses.
  * THEREFORE the unique-MAC count OVERCOUNTS people - one human can register
    as several devices and climbs as addresses rotate. Treat the count as an
    UPPER BOUND, not a person count. A real deployment needs a dedup/correction
    layer (RSSI+timing clustering, calibrated factor, payload fingerprinting,
    or a WiFi-probe cross-check).
"""

import asyncio
import csv
import sys
from datetime import datetime, timezone

from bleak import BleakScanner

csv_path = sys.argv[1] if len(sys.argv) > 1 else \
    f"ble_log_{datetime.now().strftime('%Y%m%d_%H%M%S')}.csv"

seen = {}  # mac -> {first, last, hits, rssi, name}


def now_local():
    return datetime.now(timezone.utc).astimezone()


def print_summary():
    print("\n" + "=" * 64)
    print("SCAN SUMMARY")
    print(f"  Unique MAC addresses seen : {len(seen)}")
    if seen:
        first = min(v["first"] for v in seen.values())
        secs = int((now_local() - first).total_seconds())
        print(f"  Scan duration             : {secs} sec")
        rssis = [v["rssi"] for v in seen.values() if v["rssi"] is not None]
        if rssis:
            print(f"  Strongest RSSI (closest)  : {max(rssis)} dBm")
    print(f"  CSV saved to              : {csv_path}")
    print("  Count is an UPPER BOUND (phones rotate MACs).")
    print("=" * 64)


async def main():
    f = open(csv_path, "w", newline="")
    w = csv.writer(f)
    w.writerow(["timestamp", "event", "mac", "rssi", "name", "unique_so_far"])
    f.flush()
    print(f"Logging to {csv_path}")
    print("Scanning via bleak (BlueZ/D-Bus) ... (Ctrl-C to stop)")
    print("-" * 64)
    try:
        # Default is active scanning (solicits scan responses -> more devices).
        # For a stealthier / lower-power deploy later, try
        # BleakScanner(scanning_mode="passive").
        async with BleakScanner() as scanner:
            async for dev, adv in scanner.advertisement_data():
                mac = dev.address.upper()
                rssi = adv.rssi
                name = adv.local_name or getattr(dev, "name", None) or ""
                ts = now_local()
                tstr = ts.isoformat(timespec="seconds")
                if mac in seen:
                    rec = seen[mac]
                    rec["hits"] += 1
                    rec["last"] = ts
                    if rssi is not None:
                        rec["rssi"] = rssi
                    if name:
                        rec["name"] = name
                    w.writerow([tstr, "seen", mac, rssi, rec["name"], len(seen)])
                else:
                    seen[mac] = {"first": ts, "last": ts, "hits": 1,
                                 "rssi": rssi, "name": name}
                    w.writerow([tstr, "NEW", mac, rssi, name, len(seen)])
                    f.flush()
                    rs = str(rssi) if rssi is not None else "?"
                    print(f"[{ts:%H:%M:%S}] NEW  {mac}  RSSI={rs:<4} "
                          f"{name[:20]:<20} (unique: {len(seen)})")
    finally:
        f.flush()
        f.close()
        print_summary()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        pass
